### DISCUSSION TITLE
Read Replicas | Early Access - Feedback discussion

### DISCUSSION THREAD

**Discussion by darora** [supabase/supabase#21585] [General]
This discussion is for tracking feedback and possible issues for [Supabase Read Replicas](https://supabase.com/docs/guides/platform/read-replicas).

**Comment by sbbu** [comment]
Something that feels sorely missing from the read replicas offering is a simple way to route connections. We get one postgres connection string for each replica and have to handle routing ourselves. I would love if supavisor could route reads to the nearest replica by location (to minimise latency) and writes to the primary database, all with one connection string.

The API load balancer doesn't offer this functionality either, it distributes equally among all replicas. Which makes having replicas in multiple regions seem completely pointless to me, unless I'm missing something. My team and I have found it somewhat confusing since the docs all give the impression that replicas are intended for deploying data closer to users and minimising latency, but the functionality for that isn't actually there, you have to roll it yourself.

**Reply by LordZardeck** [reply]
Yeah, we just started looking into scaling our platform, and we looked to read replicas to handle load for retrieving data, and having to route the connections on our own is a non-starter. Pretty lame that the only way to get any kind of read balancing is through the API, which we don't use.

**Reply by arthtyagi** [reply]
or, how about just look for request headers, route it yourself to the read replica?
for example, if you use drizzle:

```ts
import { NextRequest } from 'next/server';
import { getDb } from '@/lib/db';
import { usersTable } from '@/lib/schema';

export async function GET(req: NextRequest) {
  const db = getDb(req.headers);
  const users = await db.select().from(usersTable);
  return Response.json(users);
}
```

```ts
function selectReplicaByHeader(replicas: any[], headers: Headers) {
  const region = headers.get('x-region') || headers.get('x-forwarded-for');

  if (region?.includes('us-east')) return replicas[0];
  if (region?.includes('eu-west')) return replicas[1];
  return replicas[0]; // fallback
}
```
(x-region is not real as far as i know, you can get `x-vercel-id` if you're on Vercel. https://vercel.com/docs/headers/request-headers


```ts
import { withReplicas } from 'drizzle-orm/pg-core';
import { drizzle } from 'drizzle-orm/node-postgres';

export function getDb(headers: Headers) {
  const primaryDb = drizzle("postgres://.../primary");
  const read1 = drizzle("postgres://.../read1");
  const read2 = drizzle("postgres://.../read2");

  return withReplicas(primaryDb, [read1, read2], (replicas) => {
    return selectReplicaByHeader(replicas, headers);
  });
}
```

any reason something like this wouldn't help?
